File: Python/problem3.py
def largest_prime_factor(n):
    prime_factors = []
    divisor = 2
    while divisor <= n:
        if n % divisor == 0:
            prime_factors.append(divisor)
            n //= divisor
        else:
            divisor += 1
    return prime_factors[-1]


if __name__ == '__main__':
    print(largest_prime_factor(600851475143))

File: Scala/Problem3.scala
object Problem3 {
  def largestPrimeFactor(m: BigInt): BigInt = {
    @annotation.tailrec
    def f(n: BigInt, divisor: BigInt, acc: List[BigInt]): List[BigInt] =
      if (divisor > n) acc
      else if (n % divisor == 0) f(n / divisor, divisor, divisor :: acc)
      else f(n, divisor + 1, acc)
    def primeFactors(n: BigInt): List[BigInt] = f(n, 2, Nil)
    primeFactors(m).head
  }

  def main(args: Array[String]): Unit =
    println(largestPrimeFactor(600851475143L))
}
